integrated_channels/moodle/client.py

```diff
@@ -0,0 +1,233 @@
+        parsed_response = json.loads(response.text)
+        if not parsed_response.get('courses'):
+            raise ClientError('MoodleAPIClient request failed: 404 Course key '
+                              '"{}" not found in Moodle.'.format(key))
+
+        return parsed_response['courses'][0]['id']
+
+    @moodle_request_wrapper
+    def create_content_metadata(self, serialized_data):
+        """
+        The below assumes the data is dict/object.
+        Format should look like:
+        {
+          courses[0][shortname]: 'value',
+          courses[0][fullname]: 'value',
+          [...]
+          courses[1][shortname]: 'value',
+          courses[1][fullname]: 'value',
+          [...]
+        }
+        """
+        serialized_data['wsfunction'] = 'core_course_create_courses'
+        response = self._post(serialized_data)
+        return response
+
+    @moodle_request_wrapper
+    def update_content_metadata(self, serialized_data):
+        for key in list(serialized_data):
+            if 'shortname' in key:
+                moodle_course_id = self.get_course_id(serialized_data[key])
+                serialized_data[key.replace('shortname', 'id')] = moodle_course_id
+        serialized_data['wsfunction'] = 'core_course_update_courses'
+        return self._post(serialized_data)
+
+    @moodle_request_wrapper
+    def delete_content_metadata(self, serialized_data):
+        course_ids_to_delete = []
+        for key in list(serialized_data):
+            if 'shortname' in key:
+                moodle_course_id = self.get_course_id(serialized_data[key])
+                course_ids_to_delete.append(('courseids[]', moodle_course_id))
+        params = {
+            'wsfunction': 'core_course_delete_courses',
+        }
+        url = '{url}{api_path}?{querystring}'.format(
+            url=self.enterprise_configuration.moodle_base_url,
+            api_path=self.MOODLE_API_PATH,
+            querystring=urlencode(course_ids_to_delete))
+        return self._post(params, url)
+
+    def _post(self, additional_params, method_url=None):
+        """
+        Compile common params and run request's post function
+        """
+        params = {
+            'wstoken': self.token,
+            'moodlewsrestformat': 'json',
+        }
+        params.update(additional_params)
+        if method_url:
+            response = requests.post(
+                url='{url}&{querystring}'.format(
+                    url=method_url,
+                    querystring=urlencode(params)
+                )
+            )
+        else:
+            response = requests.post(
+                url='{url}{api_path}?{querystring}'.format(
+                    url=method_url if method_url else self.enterprise_configuration.moodle_base_url,
+                    api_path=self.MOODLE_API_PATH,
+                    querystring=urlencode(params)
+                )
+            )
+        return response
+
+    def _get_access_token(self):
```

Comment:
The ordering of the functions seems slightly mixed. Up to you, but we could do something similar to how we organized private/public methods in the canvas client, keep things consistent?

Reply:
I'm not 100% sure what you mean here, but assuming you mean the fact this method is defined way after it is referenced I'm not sure it really matters (I mean functionally it definitely doesn't matter, but even for stylistic purposes I'm doubtful). `_get_access_token` is used in the `__init__` method directly so there's no logical place to put it just for flow/usage purposes.

I can move it up to where the other helper functions are though if that's what you're looking for. I think its placement got messed up due to various merging while we were working on this.

Reply:
I was only looking at the order: 
`create_course_completion`,
`delete_course_completion`, 
`_get_courses`,
`get_course_id`,
`create_content_metadata`,
`update_content_metadata`,
`delete_content_metadata`,
`_post`,
`_get_access_token`
 
and wanted to just explore other options like maybe grouping the content metadata and course completion methods + grouping private methods (the canvas client has an inline comment designating where the private methods are, unnecessary sure but quite clear). If you feel that the order is already good, then definitely don't feel the need to make a change here.

Reply:
hmmm I could see ordering like:
```
_post
_get_access_token
_get_courses
get_course_id
create/update/delete_content_metadata
<whatever helper methods you create for course completion>
create/delete_course_completion
```

Does that seem like a good flow?